actionpack/lib/action_dispatch/http/url.rb

```diff
@@ -80,6 +81,7 @@ def path_for(options)
 
         def add_params(path, params)
           params = { params: params } unless params.is_a?(Hash)
+          params.transform_values! { |v| v.respond_to?(:permitted?) ? v.to_h : v }
```

Comment:
We should be doing the same as https://github.com/rails/rails/blob/2b4a9735d816d3a05c7fca207d97bebcb09c95a0/actionview/lib/action_view/helpers/url_helper.rb#L619-L628 here